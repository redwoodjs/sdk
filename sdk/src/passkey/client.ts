import {
  startRegistration,
  startAuthentication,
} from "@simplewebauthn/browser";
import {
  startPasskeyRegistration,
  finishPasskeyRegistration,
  startPasskeyLogin,
  finishPasskeyLogin,
} from "./worker.js";

export function usePasskey() {
  const register = async (username: string) => {
    // 1. Get a challenge from the worker
    const options = await startPasskeyRegistration(username);
    // 2. Ask the browser to sign the challenge
    const registration = await startRegistration(options);
    // 3. Give the signed challenge to the worker to finish the registration process
    const success = await finishPasskeyRegistration(username, registration);
    return success;
  };

  const login = async (username: string) => {
    // 1. Get a challenge from the worker
    const options = await startPasskeyLogin();
    // 2. Ask the browser to sign the challenge
    const authentication = await startAuthentication(options);
    // 3. Give the signed challenge to the worker to finish the login process
    const success = await finishPasskeyLogin(authentication);
    return success;
  };

  return { register, login };
}
