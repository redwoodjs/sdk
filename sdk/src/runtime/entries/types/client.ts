import "./shared";

export interface ClientNavigationOptions {
  onNavigate?: () => void;
  scrollToTop?: boolean;
  scrollBehavior?: "auto" | "smooth" | "instant";
}
