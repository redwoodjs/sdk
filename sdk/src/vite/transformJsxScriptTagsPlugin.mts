import {
  Project,
  Node,
  SyntaxKind,
  ImportDeclaration,
  CallExpression,
  ObjectLiteralExpression,
  SourceFile,
} from "ts-morph";
import { type Plugin, type ResolvedConfig } from "vite";
import { readFile } from "node:fs/promises";
import { pathExists } from "fs-extra";
import {
  getStylesheetsForEntryPoint,
  type StylesheetContext,
} from "./jsEntryPointsToStylesheetsPlugin.mjs";
import debug from "debug";

const log = debug("rwsdk:vite:transform-jsx-script-tags");

let manifestCache: Record<string, { file: string }> | undefined;

const readManifest = async (
  manifestPath: string,
): Promise<Record<string, { file: string }>> => {
  if (manifestCache === undefined) {
    const exists = await pathExists(manifestPath);

    if (!exists) {
      throw new Error(
        `RedwoodSDK expected client manifest to exist at ${manifestPath}. This is likely a bug. Please report it at https://github.com/redwoodjs/sdk/issues/new`,
      );
    }

    manifestCache = JSON.parse(await readFile(manifestPath, "utf-8"));
    log("Read and cached manifest from %s", manifestPath);
  }

  return manifestCache!;
};

function hasJsxFunctions(text: string): boolean {
  return (
    text.includes('jsx("script"') ||
    text.includes("jsx('script'") ||
    text.includes('jsx("link"') ||
    text.includes("jsx('link'") ||
    text.includes('jsxs("script"') ||
    text.includes("jsxs('script'") ||
    text.includes('jsxs("link"') ||
    text.includes("jsxs('link'") ||
    text.includes('jsxDEV("script"') ||
    text.includes("jsxDEV('script'") ||
    text.includes('jsxDEV("link"') ||
    text.includes("jsxDEV('link'")
  );
}

// Transform import statements in script content using ts-morph
function transformScriptImports(
  scriptContent: string,
  manifest: Record<string, any>,
): { content: string | undefined; hasChanges: boolean } {
  const scriptProject = new Project({ useInMemoryFileSystem: true });

  try {
    // Wrap in a function to make it valid JavaScript
    const wrappedContent = `function __wrapper() {${scriptContent}}`;
    const scriptFile = scriptProject.createSourceFile(
      "script.js",
      wrappedContent,
    );

    let hasChanges = false;

    // Find all CallExpressions that look like import("path")
    scriptFile
      .getDescendantsOfKind(SyntaxKind.CallExpression)
      .forEach((callExpr) => {
        const expr = callExpr.getExpression();

        // Check for both "import()" and "await import()" patterns
        const isImport = expr.getText() === "import";

        // Check for await import pattern
        const isAwaitImport =
          expr.getKind() === SyntaxKind.PropertyAccessExpression &&
          expr.getText().endsWith(".import");

        if (isImport || isAwaitImport) {
          const args = callExpr.getArguments();

          if (args.length > 0 && Node.isStringLiteral(args[0])) {
            const importPath = args[0].getLiteralValue();

            if (importPath.startsWith("/")) {
              const path = importPath.slice(1); // Remove leading slash

              if (manifest[path]) {
                const transformedPath = manifest[path].file;
                args[0].replaceWithText(`"/${transformedPath}"`);
                hasChanges = true;
              }
            }
          }
        }
      });

    if (hasChanges) {
      // Extract the transformed content from inside the wrapper function
      const fullText = scriptFile.getFullText();
      // Find content between the first { and the last }
      const startPos = fullText.indexOf("{") + 1;
      const endPos = fullText.lastIndexOf("}");
      const transformedContent = fullText.substring(startPos, endPos);

      return { content: transformedContent, hasChanges: true };
    }

    // Return the original content when no changes are made
    return { content: scriptContent, hasChanges: false };
  } catch (error) {
    // If parsing fails, fall back to the original content
    console.warn("Failed to parse inline script content:", error);
    return { content: undefined, hasChanges: false };
  }
}

function isJsxCall(callExpr: CallExpression): "jsx" | "jsxs" | "jsxDEV" | null {
  const expression = callExpr.getExpression();
  const expressionText = expression.getText();

  if (
    expressionText === "jsx" ||
    expressionText === "jsxs" ||
    expressionText === "jsxDEV"
  ) {
    return expressionText;
  }

  return null;
}

function getJsxTagName(callExpr: CallExpression): string | null {
  const args = callExpr.getArguments();
  if (args.length < 2) return null;

  const elementType = args[0];
  if (!Node.isStringLiteral(elementType)) return null;

  return elementType.getLiteralValue();
}

async function injectStylesheetLinks(
  scriptCall: CallExpression,
  props: ObjectLiteralExpression,
  context: StylesheetContext,
): Promise<boolean> {
  const srcProp = props.getProperty("src");
  if (!srcProp || !Node.isPropertyAssignment(srcProp)) {
    return false;
  }

  const srcInitializer = srcProp.getInitializer();
  if (!Node.isStringLiteral(srcInitializer)) {
    return false;
  }

  const src = srcInitializer.getLiteralValue();
  const stylesheets = await getStylesheetsForEntryPoint(src, context);

  if (stylesheets.length === 0) {
    return false;
  }

  const linkElements = stylesheets.map(
    (href) => `jsx("link", { rel: "stylesheet", href: "${href}" })`,
  );
  const scriptElement = scriptCall.getText();
  const replacement = `[${linkElements.join(", ")}, ${scriptElement}]`;

  scriptCall.replaceWithText(replacement);
  return true;
}

function findRwsdkWorkerImport(sourceFile: SourceFile): {
  sdkWorkerImportDecl: ImportDeclaration | undefined;
  hasRequestInfoImport: boolean;
} {
  let sdkWorkerImportDecl: ImportDeclaration | undefined;
  let hasRequestInfoImport = false;

  sourceFile.getImportDeclarations().forEach((importDecl) => {
    const moduleSpecifier = importDecl.getModuleSpecifierValue();
    if (moduleSpecifier === "rwsdk/worker") {
      sdkWorkerImportDecl = importDecl;
      if (
        importDecl
          .getNamedImports()
          .some((namedImport) => namedImport.getName() === "requestInfo")
      ) {
        hasRequestInfoImport = true;
      }
    }
  });

  return { sdkWorkerImportDecl, hasRequestInfoImport };
}

function addRequestInfoImport(
  sourceFile: SourceFile,
  sdkWorkerImportDecl: ImportDeclaration | undefined,
) {
  if (sdkWorkerImportDecl) {
    // Module is imported but need to add requestInfo
    sdkWorkerImportDecl.addNamedImport("requestInfo");
  } else {
    // Add new import declaration
    sourceFile.addImportDeclaration({
      moduleSpecifier: "rwsdk/worker",
      namedImports: ["requestInfo"],
    });
  }
}

async function transformScriptTag(
  callExpr: CallExpression,
  props: ObjectLiteralExpression,
  manifest: Record<string, any>,
  context: StylesheetContext,
) {
  let hasModifications = false;
  const stylesheetInjected = await injectStylesheetLinks(
    callExpr,
    props,
    context,
  );

  if (stylesheetInjected) {
    // The original callExpr was replaced, we can't do more work on its props.
    // The props would need to be re-parsed from the new fragment's script element.
    // For now, we assume this is the only transformation needed for this node.
    // This is a limitation: we won't add a nonce or transform the src of a script
    // that also had stylesheets injected.
    return true;
  }

  const properties = props.getProperties();

  // We loop again because the props might have changed
  for (const prop of properties) {
    if (Node.isPropertyAssignment(prop)) {
      const propName = prop.getName();
      const initializer = prop.getInitializer();

      // Transform src for manifest
      if (
        propName === "src" &&
        (Node.isStringLiteral(initializer) ||
          Node.isNoSubstitutionTemplateLiteral(initializer))
      ) {
        const srcValue = initializer.getLiteralValue();
        if (srcValue.startsWith("/") && manifest[srcValue.slice(1)]) {
          const path = srcValue.slice(1);
          const transformedSrc = manifest[path].file;
          const quote = Node.isNoSubstitutionTemplateLiteral(initializer)
            ? "`"
            : initializer.getText().charAt(0);
          initializer.replaceWithText(`${quote}/${transformedSrc}${quote}`);
          hasModifications = true;
        }
      }

      // Transform inline script children
      if (
        propName === "children" &&
        (Node.isStringLiteral(initializer) ||
          Node.isNoSubstitutionTemplateLiteral(initializer))
      ) {
        const scriptContent = initializer.getLiteralValue();
        const { content: transformedContent, hasChanges } =
          transformScriptImports(scriptContent, manifest);
        if (hasChanges && transformedContent) {
          const quote = Node.isNoSubstitutionTemplateLiteral(initializer)
            ? ""
            : '"';
          const replacement = Node.isNoSubstitutionTemplateLiteral(initializer)
            ? "`" + transformedContent + "`"
            : JSON.stringify(transformedContent);
          initializer.replaceWithText(replacement);
          hasModifications = true;
        }
      }
    }
  }

  return hasModifications;
}

function shouldScriptHaveNonce(props: ObjectLiteralExpression): boolean {
  let hasDangerouslySetInnerHTML = false;
  let hasNonce = false;
  let hasChildren = false;
  let hasSrc = false;

  for (const prop of props.getProperties()) {
    if (Node.isPropertyAssignment(prop)) {
      const propName = prop.getName();
      if (propName === "nonce") hasNonce = true;
      if (propName === "dangerouslySetInnerHTML")
        hasDangerouslySetInnerHTML = true;
      if (propName === "children") hasChildren = true;
      if (propName === "src") hasSrc = true;
    }
  }

  return !hasNonce && !hasDangerouslySetInnerHTML && (hasChildren || hasSrc);
}

function injectNonces(nodes: ObjectLiteralExpression[]) {
  for (const props of nodes) {
    props.addPropertyAssignment({
      name: "nonce",
      initializer: "requestInfo.rw.nonce",
    });
  }
}

function transformLinkTag(
  props: ObjectLiteralExpression,
  manifest: Record<string, any>,
) {
  let isPreload = false;
  let hrefValue = null;
  let hrefInitializer: Node | undefined;

  for (const prop of props.getProperties()) {
    if (Node.isPropertyAssignment(prop)) {
      const propName = prop.getName();
      const initializer = prop.getInitializer();
      if (
        propName === "rel" &&
        initializer &&
        Node.isStringLiteral(initializer)
      ) {
        const relValue = initializer.getLiteralValue();
        if (relValue === "preload" || relValue === "modulepreload") {
          isPreload = true;
        }
      }
      if (
        propName === "href" &&
        initializer &&
        Node.isStringLiteral(initializer)
      ) {
        hrefInitializer = initializer;
        hrefValue = initializer.getLiteralValue();
      }
    }
  }

  if (
    isPreload &&
    hrefValue &&
    hrefInitializer &&
    hrefValue.startsWith("/") &&
    manifest[hrefValue.slice(1)]
  ) {
    const path = hrefValue.slice(1);
    const transformedHref = manifest[path].file;
    const quote = (hrefInitializer as any).getText().charAt(0);
    (hrefInitializer as any).replaceWithText(
      `${quote}/${transformedHref}${quote}`,
    );
    return true;
  }

  return false;
}

export async function transformJsxScriptTagsCode(
  code: string,
  manifest: Record<string, any> = {},
  context: StylesheetContext,
) {
  // context(justinvdm, 15 Jun 2025): Optimization to exit early
  // to avoidunnecessary ts-morph parsing
  if (!hasJsxFunctions(code)) {
    return;
  }

  const project = new Project({ useInMemoryFileSystem: true });
  const sourceFile = project.createSourceFile("temp.tsx", code);

  let hasModifications = false;
  const { sdkWorkerImportDecl, hasRequestInfoImport } =
    findRwsdkWorkerImport(sourceFile);

  const scriptsNeedingNonce: ObjectLiteralExpression[] = [];
  const callExpressions = sourceFile.getDescendantsOfKind(
    SyntaxKind.CallExpression,
  );

  // Look for jsx function calls (jsx, jsxs, jsxDEV)
  for (const callExpr of callExpressions) {
    if (!isJsxCall(callExpr)) {
      continue;
    }

    const tagName = getJsxTagName(callExpr);
    if (!tagName) {
      continue;
    }

    const props = callExpr.getArguments()[1];
    if (!Node.isObjectLiteralExpression(props)) {
      continue;
    }

    if (tagName === "script") {
      const modified = await transformScriptTag(
        callExpr,
        props,
        manifest,
        context,
      );
      if (modified) {
        hasModifications = true;
        // If the node was replaced (e.g., by stylesheet injection),
        // we should skip the rest of the logic for this node.
        if (callExpr.wasForgotten()) {
          continue;
        }
      }

      if (shouldScriptHaveNonce(props)) {
        scriptsNeedingNonce.push(props);
      }
    } else if (tagName === "link") {
      const modified = transformLinkTag(props, manifest);
      if (modified) hasModifications = true;
    }
  }

  let needsRequestInfoImport = false;
  if (scriptsNeedingNonce.length > 0) {
    injectNonces(scriptsNeedingNonce);
    hasModifications = true;
    if (!hasRequestInfoImport) {
      needsRequestInfoImport = true;
    }
  }

  // Add requestInfo import if needed and not already imported
  if (needsRequestInfoImport) {
    addRequestInfoImport(sourceFile, sdkWorkerImportDecl);
  }

  // Return the transformed code only if modifications were made
  if (hasModifications) {
    return {
      code: sourceFile.getFullText(),
      map: null,
    };
  }

  return;
}

export const transformJsxScriptTagsPlugin = ({
  manifestPath,
}: {
  manifestPath: string;
}): Plugin => {
  let isBuild = false;
  let config: ResolvedConfig;

  return {
    name: "rwsdk:transform-jsx-script-tags",

    configResolved(resolvedConfig) {
      isBuild = resolvedConfig.command === "build";
      config = resolvedConfig;
    },

    async transform(code) {
      if (this.environment.name !== "worker") {
        return;
      }

      const manifest = isBuild ? await readManifest(manifestPath) : {};
      const context: StylesheetContext = {
        isBuild,
        projectRootDir: config.root,
        buildOutDir: config.build.outDir,
      };

      return transformJsxScriptTagsCode(code, manifest, context);
    },
  };
};
