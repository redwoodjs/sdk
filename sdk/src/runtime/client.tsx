import { clientWebpackRequire } from "./imports/client";
import { type CallServerCallback } from "react-server-dom-webpack/client.browser";

// NOTE: `react-server-dom-webpack` uses this global to load modules,
// so we need to define it here before importing "react-server-dom-webpack."
globalThis.__webpack_require__ = clientWebpackRequire;

export type ActionResponse<Result> = {
  node: React.ReactNode;
  actionResult: Result;
};

type TransportContext = {
  setRscPayload: <Result>(v: Promise<ActionResponse<Result>>) => void;
};

export type Transport = (context: TransportContext) => CallServerCallback;

export type CreateCallServer = (
  context: TransportContext,
) => <Result>(id: null | string, args: null | unknown[]) => Promise<Result>;

export const fetchTransport: Transport = (transportContext) => {
  const fetchCallServer = async <Result,>(
    id: null | string,
    args: null | unknown[],
  ): Promise<Result> => {
    const { createFromFetch, encodeReply } = await import(
      "react-server-dom-webpack/client.browser"
    );

    const url = new URL(window.location.href);
    url.searchParams.set("__rsc", "");

    if (id != null) {
      url.searchParams.set("__rsc_action_id", id);
    }

    const streamData = createFromFetch(
      fetch(url, {
        method: "POST",
        body: args != null ? await encodeReply(args) : null,
      }),
      { callServer: fetchCallServer },
    ) as Promise<ActionResponse<Result>>;

    transportContext.setRscPayload(streamData);
    const result = await streamData;
    return (result as { actionResult: Result }).actionResult;
  };

  return fetchCallServer;
};

export const initClient = async ({
  transport = fetchTransport,
}: {
  transport?: Transport;
} = {}) => {
  const transportContext: TransportContext = {
    setRscPayload: () => {},
  };

  let transportCallServer = transport(transportContext);

  const callServer = (id: any, args: any) => transportCallServer(id, args);

  const upgradeToRealtime = async ({ key }: { key?: string } = {}) => {
    const { realtimeTransport } = await import("./lib/realtime/client");
    const createRealtimeTransport = realtimeTransport({ key });
    transportCallServer = createRealtimeTransport(transportContext);
  };

  globalThis.__rsc_callServer = callServer;

  globalThis.__rw = {
    callServer,
    upgradeToRealtime,
  };

  const rootEl = document.getElementById("root");

  if (!rootEl) {
    throw new Error('no element with id "root"');
  }

  const React = await import("react");
  const { hydrateRoot } = await import("react-dom/client");
  // @ts-ignore: todo(peterp, 2024-11-27): Type these properly.
  const { createFromReadableStream, createFromFetch, encodeReply } =
    await import("react-server-dom-webpack/client.browser");
  const { rscStream } = await import("rsc-html-stream/client");

  let rscPayload: any;
  rscPayload ??= createFromReadableStream(rscStream, {
    callServer,
  });

  function Content() {
    const [streamData, setStreamData] = React.useState(rscPayload);
    const [_isPending, startTransition] = React.useTransition();
    transportContext.setRscPayload = (v) =>
      startTransition(() => setStreamData(v));
    return <>{React.use<{ node: React.ReactNode }>(streamData).node}</>;
  }

  hydrateRoot(rootEl, <Content />);

  if (import.meta.hot) {
    import.meta.hot.on("rsc:update", (e) => {
      console.log("[rw-sdk] hot update", e.file);
      callServer("__rsc_hot_update", [e.file]);
    });
  }
};
