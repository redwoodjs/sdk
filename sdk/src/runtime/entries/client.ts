import "./types/client";
export * from "../client/client";
export * from "../client/navigation";
export * from "../lib/streams/consumeEventStream";
export * from "../register/client";
