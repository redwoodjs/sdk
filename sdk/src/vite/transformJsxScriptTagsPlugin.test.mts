import { describe, it, expect } from "vitest";
import { transformJsxScriptTagsCode } from "./transformJsxScriptTagsPlugin.mjs";

describe("transformJsxScriptTagsCode", () => {
  const mockManifest = {
    "src/client.tsx": { file: "assets/client-a1b2c3d4.js" },
    "src/entry.js": { file: "assets/entry-e5f6g7h8.js" },
    "src/styles.css": { file: "assets/styles-i9j0k1l2.css" },
  };

  it("transforms script src attributes in JSX", async () => {
    const code = `
      jsx("script", {
        src: "/src/client.tsx",
        type: "module"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/assets/client-a1b2c3d4.js",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("transforms inline scripts with dynamic imports", async () => {
    const code = `
      jsx("script", {
        type: "module",
        children: "import('/src/client.tsx').then(module => { console.log(module); })"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: "import(\\"\/assets\/client-a1b2c3d4.js\\").then(module => { console.log(module); })",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("transforms inline scripts with type=module", async () => {
    const code = `
      jsx("script", { type: "module", children: "import('/src/client.tsx')" })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", { type: "module", children: "import(\\"\/assets\/client-a1b2c3d4.js\\")",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("transforms inline scripts with multiline content", async () => {
    const code = `
      jsx("script", {
        type: "module",
        children: \`
          // Some comments here
          const init = async () => {
            await import('/src/entry.js');
            console.log('initialized');
          };
          init();
        \`
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: \`
          // Some comments here
          const init = async () => {
            await import("/assets/entry-e5f6g7h8.js");
            console.log('initialized');
          };
          init();
        \`,
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("transforms multiple imports in the same inline script", async () => {
    const code = `
      jsx("script", {
        type: "module",
        children: \`
          import('/src/client.tsx');
          import('/src/entry.js');
        \`
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: \`
          import("/assets/client-a1b2c3d4.js");
          import("/assets/entry-e5f6g7h8.js");
        \`,
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("transforms link href attributes with preload rel", async () => {
    const code = `
      jsx("link", {
        rel: "preload",
        href: "/src/client.tsx",
        as: "script"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`
      jsx("link", {
        rel: "preload",
        href: "/assets/client-a1b2c3d4.js",
        as: "script"
      })
    `);
  });

  it("transforms link href attributes with modulepreload rel", async () => {
    const code = `
      jsx("link", {
        href: "/src/client.tsx",
        rel: "modulepreload"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`
      jsx("link", {
        href: "/assets/client-a1b2c3d4.js",
        rel: "modulepreload"
      })
    `);
  });

  it("transforms real-world Document component example", async () => {
    const code = `
      jsx("html", {
        lang: "en",
        children: [
          jsx("head", {
            children: [
              jsx("meta", { charSet: "utf-8" }),
              jsx("meta", { name: "viewport", content: "width=device-width, initial-scale=1" }),
              jsx("title", { children: "@redwoodjs/starter-standard" }),
              jsx("link", { rel: "modulepreload", href: "/src/client.tsx", as: "script" })
            ]
          }),
          jsx("body", {
            children: [
              jsx("div", { id: "root", children: props.children }),
              jsx("script", { children: 'import("/src/client.tsx")' })
            ]
          })
        ]
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("html", {
        lang: "en",
        children: [
          jsx("head", {
            children: [
              jsx("meta", { charSet: "utf-8" }),
              jsx("meta", { name: "viewport", content: "width=device-width, initial-scale=1" }),
              jsx("title", { children: "@redwoodjs/starter-standard" }),
              jsx("link", { rel: "modulepreload", href: "/assets/client-a1b2c3d4.js", as: "script" })
            ]
          }),
          jsx("body", {
            children: [
              jsx("div", { id: "root", children: props.children }),
              jsx("script", { children: "import(\\"\/assets\/client-a1b2c3d4.js\\")",
                  nonce: requestInfo.rw.nonce
            })
            ]
          })
        ]
      })
    `);
  });

  it("returns null when no transformations are needed", async () => {
    const code = `
      jsx("div", { children: "No scripts or links here" })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result).toBeUndefined();
  });

  it("handles paths not found in manifest", async () => {
    const code = `
      jsx("script", {
        src: "/src/non-existent.js",
        type: "module"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/src/non-existent.js",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("adds nonce to script tags with src attribute and imports requestInfo", async () => {
    const code = `
      jsx("script", {
        src: "/src/client.tsx",
        type: "module"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, mockManifest);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/assets/client-a1b2c3d4.js",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("adds nonce to script tags with string literal children", async () => {
    const code = `
      jsx("script", {
        type: "module",
        children: "console.log('hello world')"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, {});

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("does not add nonce to script tags with dangerouslySetInnerHTML", async () => {
    const code = `
      jsx("script", {
        type: "module",
        dangerouslySetInnerHTML: { __html: "console.log('hello world')" }
      })
    `;

    const result = await transformJsxScriptTagsCode(code, {});

    expect(result?.code).toEqual(undefined);
  });

  it("does not add nonce to script tags that already have nonce", async () => {
    const code = `
      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
        nonce: "existing-nonce"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, {});

    expect(result?.code).toEqual(undefined);
  });

  it("uses existing requestInfo import if already present", async () => {
    const code = `
      import { foo } from 'bar';
      import { requestInfo, someOtherThing } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, {});

    expect(result?.code).toEqual(`
      import { foo } from 'bar';
      import { requestInfo, someOtherThing } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    `);

    // Ensure we didn't duplicate the import
    const importCount = (result?.code.match(/from "rwsdk\/worker"/g) || [])
      .length;
    expect(importCount).toBe(1);
  });

  it("adds requestInfo to existing SDK import if module already imported", async () => {
    const code = `
      import { foo } from 'bar';
      import { someOtherThing } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')"
      })
    `;

    const result = await transformJsxScriptTagsCode(code, {});

    expect(result?.code).toEqual(`
      import { foo } from 'bar';
      import { someOtherThing, requestInfo } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    `);
  });

  it("works in development mode without a manifest", async () => {
    const code = `
      jsx("script", {
        src: "/src/client.tsx",
        type: "module"
      })
    `;

    // Call without providing manifest (simulating dev mode)
    const result = await transformJsxScriptTagsCode(code);

    expect(result?.code).toEqual(`import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/src/client.tsx",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    `);
  });
});
