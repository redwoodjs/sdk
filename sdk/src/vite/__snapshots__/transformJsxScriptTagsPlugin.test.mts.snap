// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformJsxScriptTagsCode > adds nonce to script tags with src attribute and imports requestInfo 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/src/client.tsx",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > adds nonce to script tags with string literal children 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > adds requestInfo to existing SDK import if module already imported 1`] = `
"
      import { foo } from 'bar';
      import { someOtherThing, requestInfo } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > handles paths not found in manifest 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/src/non-existent.js",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > injects stylesheet links for scripts with src 1`] = `
"
      import { jsx } from 'react/jsx-runtime';
import { requestInfo } from "rwsdk/worker";

      function Document() {
        return jsx("body", { children: [
          jsx("script", { src: "/src/client.tsx",
              nonce: requestInfo.rw.nonce
        })
        ]});
      }
    "
`;

exports[`transformJsxScriptTagsCode > transforms inline scripts with dynamic imports 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: "import(\\"/assets/client-a1b2c3d4.js\\").then(module => { console.log(module); })",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > transforms inline scripts with multiline content 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: \`
          // Some comments here
          const init = async () => {
            await import("/assets/entry-e5f6g7h8.js");
            console.log('initialized');
          };
          init();
        \`,
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > transforms inline scripts with type=module 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", { type: "module", children: "import(\\"/assets/client-a1b2c3d4.js\\")",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > transforms link href attributes with modulepreload rel 1`] = `
"
      jsx("link", {
        href: "/assets/client-a1b2c3d4.js",
        rel: "modulepreload"
      })
    "
`;

exports[`transformJsxScriptTagsCode > transforms link href attributes with preload rel 1`] = `
"
      jsx("link", {
        rel: "preload",
        href: "/assets/client-a1b2c3d4.js",
        as: "script"
      })
    "
`;

exports[`transformJsxScriptTagsCode > transforms multiple imports in the same inline script 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        type: "module",
        children: \`
          import("/assets/client-a1b2c3d4.js");
          import("/assets/entry-e5f6g7h8.js");
        \`,
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > transforms real-world Document component example 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("html", {
        lang: "en",
        children: [
          jsx("head", {
            children: [
              jsx("meta", { charSet: "utf-8" }),
              jsx("meta", { name: "viewport", content: "width=device-width, initial-scale=1" }),
              jsx("title", { children: "@redwoodjs/starter-standard" }),
              jsx("link", { rel: "modulepreload", href: "/assets/client-a1b2c3d4.js", as: "script" })
            ]
          }),
          jsx("body", {
            children: [
              jsx("div", { id: "root", children: props.children }),
              jsx("script", { children: "import(\\"/assets/client-a1b2c3d4.js\\")",
                  nonce: requestInfo.rw.nonce
            })
            ]
          })
        ]
      })
    "
`;

exports[`transformJsxScriptTagsCode > transforms script src attributes in JSX 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/assets/client-a1b2c3d4.js",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > uses existing requestInfo import if already present 1`] = `
"
      import { foo } from 'bar';
      import { requestInfo, someOtherThing } from "rwsdk/worker";
      
      jsx("script", {
        type: "module",
        children: "console.log('hello world')",
          nonce: requestInfo.rw.nonce
    })
    "
`;

exports[`transformJsxScriptTagsCode > works in development mode without a manifest 1`] = `
"import { requestInfo } from "rwsdk/worker";

      jsx("script", {
        src: "/src/client.tsx",
        type: "module",
          nonce: requestInfo.rw.nonce
    })
    "
`;
